signalr-go-server/pkg/signalr/server.go

```diff
@@ -0,0 +1,265 @@
+	defaultHubClients := defaultHubClients{
+		lifetimeManager: &lifetimeManager,
+		allCache:        allClientProxy{lifetimeManager: &lifetimeManager},
+	}
+
+	groupManager := defaultGroupManager{
+		lifetimeManager: &lifetimeManager,
+	}
+
+	hubType := reflect.TypeOf(hubPrototype)
+
+	mux.HandleFunc(fmt.Sprintf("%s/negotiate", path), negotiateHandler)
+	mux.Handle(path, websocket.Handler(func(ws *websocket.Conn) {
+		connectionID := ws.Request().URL.Query().Get("id")
+
+		if len(connectionID) == 0 {
+			// Support websocket connection without negotiate
+			connectionID = getConnectionID()
+		}
+
+		// Copy hubPrototype
+		hub := reflect.New(reflect.ValueOf(hubPrototype).Elem().Type()).Interface().(HubInterface)
+
+		hubContext := &defaultHubContext{
+			clients: &contextHubClients{
+				defaultHubClients: &defaultHubClients,
+				connectionID:      connectionID,
+			},
+			groups: &groupManager,
+		}
+
+		hub.Initialize(hubContext)
+
+		hubInfo := &hubInfo{
+			hub:             hub,
+			lifetimeManager: &lifetimeManager,
+			methods:         make(map[string]reflect.Value),
+		}
+
+		hubValue := reflect.ValueOf(hub)
+		for i := 0; i < hubType.NumMethod(); i++ {
+			m := hubType.Method(i)
+			hubInfo.methods[strings.ToLower(m.Name)] = hubValue.Method(i)
+		}
+
+		hubConnectionHandler(connectionID, ws, hubInfo)
+	}))
+}
+
+// Implementation
+
+type hubInfo struct {
+	hub             HubInterface
+	lifetimeManager HubLifetimeManager
+	methods         map[string]reflect.Value
+}
+
+func hubConnectionHandler(connectionID string, ws *websocket.Conn, hubInfo *hubInfo) {
+	var err error
+	var buf bytes.Buffer
+	protocol, err := processHandshake(ws, &buf)
+
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	conn := webSocketHubConnection{protocol: protocol, connectionID: connectionID, ws: ws}
+	conn.start()
+
+	streamer := newStreamer(conn)
+	streamClient := newStreamClient()
+
+	hubInfo.lifetimeManager.OnConnected(&conn)
+	hubInfo.hub.OnConnected(connectionID)
+
+	// Start sending pings to the client
+	pings := startPingClientLoop(conn)
+
+	for conn.isConnected() {
+		for {
+			message, err := protocol.ReadMessage(&buf)
+
+			if err != nil {
+				// Partial message, need more data
+				break
+			}
+
+			switch message.(type) {
+			case invocationMessage:
+				invocation := message.(invocationMessage)
+				// Dispatch invocation here
+				if method, ok := hubInfo.methods[strings.ToLower(invocation.Target)]; !ok {
+					// Unable to find the method
+					conn.completion(invocation.InvocationID, nil, fmt.Sprintf("Unknown method %s", invocation.Target))
+				} else if in, clientStreaming, err := buildMethodArguments(method, invocation, streamClient, protocol); err != nil {
+					// argument build failed
+					conn.completion(invocation.InvocationID, nil, err.Error())
+				} else if clientStreaming {
+					// let the receiving method run idependently
+					go func() {
+						defer func() {
+							if err := recover(); err != nil {
```

Comment:
What's recover()?

Reply:
deferred functions are called after the surrounding block ends, even if the block panicked. recover() then returns the panics error and stack